tools/db_bench_tool.cc

```diff
@@ -6776,15 +6790,63 @@ class Benchmark {
     Duration duration(FLAGS_duration, readwrites_);
 
     std::unique_ptr<const char[]> key_guard;
+    std::vector<std::string> inserted_keys_sample;
+    std::default_random_engine bergen;
+    double p = 0.0;
+    // Create Bernoulli distribution parameter for number
+    // of overwrites if user specified num_overwrites.
+    if (FLAGS_num_overwrites >= 0) {
+      p = FLAGS_num_overwrites * 1.0 / FLAGS_num;
+      p = p > 1.0 ? 1.0 : p;
+    }
+    std::bernoulli_distribution distribution(p);
     Slice key = AllocateKey(&key_guard);
+    char* start = const_cast<char*>(key.data());
+    char* pos = start;
     std::unique_ptr<char[]> ts_guard;
     if (user_timestamp_size_ > 0) {
       ts_guard.reset(new char[user_timestamp_size_]);
     }
     // the number of iterations is the larger of read_ or write_
     while (!duration.Done(1)) {
       DB* db = SelectDB(thread);
-      GenerateKeyFromInt(thread->rand.Next() % FLAGS_num, FLAGS_num, &key);
+      // If user specified overwrite ratio:
+      if (FLAGS_num_overwrites >= 0) {
+        // Decide if we do an overwrite (probability p).
+        // Note: an overwrite is possible only once at least
+        // one key has been inserted.
+        if (inserted_keys_sample.size() > 0 && distribution(bergen)) {
+          // If overwrite, randomly select key from sample (reservoir).
+          std::string newkey =
+              inserted_keys_sample[thread->rand.Next() %
+                                   inserted_keys_sample.size()];
+          // Copy string chars to key.
+          for (size_t i = 0; i < key.size(); i++) {
+            pos[i] = newkey[i];
+          }
+        } else {
+          // We perform a pure write.
+          Status sget = Status::OK();
+          // Keep generating new key until we obtain a key
+          // not already present in the DB.
+          do {
```

Comment:
It seems like in some cases this could spin and quietly dominate the cost of the test, leading to misleading results. Thoughts?

Reply:
After a bit of testing , it looks like there is a substantial difference in throughput for different `overwrite_probabilities`. The main driver of the difference is actually the `db->Get` operation that checks if the newly generated key is existing or not (even when the do-while loop is replaced by a single iteration).
In my opinion, the only type of benchmark that wouldn't be affected is the `fillSeq` benchmarl - because then we know exactly which keys have been inserted before or not (so no need for `db->Get`), and there is no need for random key generation either (except maybe a single call for selecting a random overwrite).